### ISSUE TITLE
Fix ranking issue

### ISSUE THREAD

**Issue by varunKT001** [opencodeiiita/Leaderboard-Frontend#37]
Several times scores are ordered incorrectly, like the high score comes below the low score. 
Here is a screenshot regarding the same : 
![image](https://user-images.githubusercontent.com/83509023/136015467-829944d0-2eb8-40cf-974d-a3bd1d0128b7.png)

**Comment by shlesh** [comment]
claim

**Comment by varunKT001** [comment]
@infern018 i have also fixed it

**Comment by shlesh** [comment]
> 
> 
> @infern018 i have also fixed it

gg😌🔥🔥🔥🔥🔥🔥

**Comment by batflarrow** [comment]
The issue is not with the order but with the API being called even if one call is made already. A better fix will be to only fetch the next results after the previous results have been fetched and rendered. The issue is that when scrolling down we have to just add a check if a current request is being made then not make another request. But currently what happening is multiple requests are being made because of which the ordering is getting a little bit shuffled.
A single boolean variable isFetching will clear out this issue. I wrote the code but for it already but forgot to push it in my pr so this one's on me.

**Comment by varunKT001** [comment]
> The issue is not with the order but with the API being called even if one call is made already. A better fix will be to only fetch the next results after the previous results have been fetched and rendered. The issue is that when scrolling down we have to just add a check if a current request is being made then not make another request. But currently what happening is multiple requests are being made because of which the ordering is getting a little bit shuffled. A single boolean variable isFetching will clear out this issue. I wrote the code for it already but forgot to push it in my pr so this one's on me.

In my opinion it should not be the case. Whenever the getData() function will be called by the scroll event listner, it will go in the queue, and will begin its execution after the call stack has been empty. Therefore the new results are only fetched when the previous results are rendered.

**Comment by batflarrow** [comment]
> > The issue is not with the order but with the API being called even if one call is made already. A better fix will be to only fetch the next results after the previous results have been fetched and rendered. The issue is that when scrolling down we have to just add a check if a current request is being made then not make another request. But currently what happening is multiple requests are being made because of which the ordering is getting a little bit shuffled. A single boolean variable isFetching will clear out this issue. I wrote the code for it already but forgot to push it in my pr so this one's on me.
> 
> In my opinion it should not be the case. Whenever the getData() function will be called by the scroll event listner, it will go in the queue, and will begin its execution after the call stack has been empty. Therefore the new results are only fetched when the previous results are rendered.

But currently, getData() function was being called multiple times and it being an asynchronous function it was running in parallel, and each getData() call was not ending after the same period of time so if the getData() function is called simultaneosly for page=4 and page=5 it may happen that the results with the page=5 will be fetched before the results with page =4 so the results on page=5 were getting rendered before the page=4 and hence the clutter. I fixed this in my pr and it's working fine now.

**Comment by varunKT001** [comment]
> > > The issue is not with the order but with the API being called even if one call is made already. A better fix will be to only fetch the next results after the previous results have been fetched and rendered. The issue is that when scrolling down we have to just add a check if a current request is being made then not make another request. But currently what happening is multiple requests are being made because of which the ordering is getting a little bit shuffled. A single boolean variable isFetching will clear out this issue. I wrote the code for it already but forgot to push it in my pr so this one's on me.
> > 
> > 
> > In my opinion it should not be the case. Whenever the getData() function will be called by the scroll event listner, it will go in the queue, and will begin its execution after the call stack has been empty. Therefore the new results are only fetched when the previous results are rendered.
> 
> But currently, getData() function was being called multiple times and it is an asynchronous function it was running in parallel, and each getData() call was not ending after the same period of time so if the getData() function is called simultaneously for page=4 and page=5 it may happen that the results with the page=5 will be fetched before the results with page =4 so the results on page=5 were getting rendered before the page=4 and hence the clutter. Yes, getData() function will move back to call Stack once it gets executed but here multiple getData() were being pushed onto the call stack simultaneously on reaching the end of the page. I fixed this in my pr by only calling the getData() if one getData() is already not in the queue. I had implemented this earlier itself but somehow forgot to push it in my pr and it's working fine now. https://batflarrow.github.io/Leaderboard-Frontend/
> 
> ![image](https://user-images.githubusercontent.com/56157504/136036170-d8dbbb6e-0fca-432b-9ba8-d4685d09470a.png)
> 
> This is the screenshot of the network tab of the currently deployed version. You can see the call to page =3 and page =4 are called simultaneously but API with page =4 resolves before the API with page =3 hence getData() comes out of event loop and the data with page=4 gets rendered before the data with page=3 and hence the clutter.

Yeah, true. but JS is single threaded and even if multiple getData() gets called, it wont execute simultaneouly. 
For example : refer to this codepen I made [codepen](https://codepen.io/varuntiwari2001/pen/yLXdKVd)
On every scroll the getData() will be called and for every promise returned by getData(), addToTable() will be called. If users scrolls the getData() function calls will be queued. The queued functions will only resume after the call stack is empty.

**Comment by batflarrow** [comment]
> > > > The issue is not with the order but with the API being called even if one call is made already. A better fix will be to only fetch the next results after the previous results have been fetched and rendered. The issue is that when scrolling down we have to just add a check if a current request is being made then not make another request. But currently what happening is multiple requests are being made because of which the ordering is getting a little bit shuffled. A single boolean variable isFetching will clear out this issue. I wrote the code for it already but forgot to push it in my pr so this one's on me.
> > > 
> > > 
> > > In my opinion it should not be the case. Whenever the getData() function will be called by the scroll event listner, it will go in the queue, and will begin its execution after the call stack has been empty. Therefore the new results are only fetched when the previous results are rendered.
> > 
> > 
> > But currently, getData() function was being called multiple times and it is an asynchronous function it was running in parallel, and each getData() call was not ending after the same period of time so if the getData() function is called simultaneously for page=4 and page=5 it may happen that the results with the page=5 will be fetched before the results with page =4 so the results on page=5 were getting rendered before the page=4 and hence the clutter. Yes, getData() function will move back to call Stack once it gets executed but here multiple getData() were being pushed onto the call stack simultaneously on reaching the end of the page. I fixed this in my pr by only calling the getData() if one getData() is already not in the queue. I had implemented this earlier itself but somehow forgot to push it in my pr and it's working fine now. https://batflarrow.github.io/Leaderboard-Frontend/
> > ![image](https://user-images.githubusercontent.com/56157504/136036170-d8dbbb6e-0fca-432b-9ba8-d4685d09470a.png)
> > This is the screenshot of the network tab of the currently deployed version. You can see the call to page =3 and page =4 are called simultaneously but API with page =4 resolves before the API with page =3 hence getData() comes out of event loop and the data with page=4 gets rendered before the data with page=3 and hence the clutter.
> 
> Yeah, true. but JS is single threaded and even if multiple getData() gets called, it wont execute simultaneouly. For example : refer to this codepen I made [codepen](https://codepen.io/varuntiwari2001/pen/yLXdKVd) On every scroll the getData() will be called and for every promise returned by getData(), addToTable() will be called. If users scrolls the getData() function calls will be queued. The queued functions will only resume after the call stack is empty. The getData() functions in the queue will be in order of page number, and for each data set recieved, the addToTable() will run, before the stack becomes empty. So i dont think that's a problem 🤔. I may be wrong. Pls correct me if making any mistake 😊

I edited the [codepen](https://codepen.io/batflarrow/pen/gORNzaE?editors=1011) slightly to introduce time lag like the apis each time getData() is called it waits for less amount of time to resolve so if getData() was called for the first time it will wait for 5 seconds if it was called for the second time it will wait for 4 seconds and so on and also after resolving I send the time value when the function was called and sent it to the addToPage() function. So according to you if 2-3 getData() are called simultaneously the function with time = 5 should be called first and then time =4 and then time =3 and then the div's should be printed in the order with time =5 then time =4 then time =3 but you can clearly see that's not happening, in the codepen. 
![image](https://user-images.githubusercontent.com/56157504/136052449-d5443ec6-5d2a-4105-ae65-b03f448fbb67.png)

**Comment by batflarrow** [comment]
> > > > > The issue is not with the order but with the API being called even if one call is made already. A better fix will be to only fetch the next results after the previous results have been fetched and rendered. The issue is that when scrolling down we have to just add a check if a current request is being made then not make another request. But currently what happening is multiple requests are being made because of which the ordering is getting a little bit shuffled. A single boolean variable isFetching will clear out this issue. I wrote the code for it already but forgot to push it in my pr so this one's on me.
> > > > 
> > > > 
> > > > In my opinion it should not be the case. Whenever the getData() function will be called by the scroll event listner, it will go in the queue, and will begin its execution after the call stack has been empty. Therefore the new results are only fetched when the previous results are rendered.
> > > 
> > > 
> > > But currently, getData() function was being called multiple times and it is an asynchronous function it was running in parallel, and each getData() call was not ending after the same period of time so if the getData() function is called simultaneously for page=4 and page=5 it may happen that the results with the page=5 will be fetched before the results with page =4 so the results on page=5 were getting rendered before the page=4 and hence the clutter. Yes, getData() function will move back to call Stack once it gets executed but here multiple getData() were being pushed onto the call stack simultaneously on reaching the end of the page. I fixed this in my pr by only calling the getData() if one getData() is already not in the queue. I had implemented this earlier itself but somehow forgot to push it in my pr and it's working fine now. https://batflarrow.github.io/Leaderboard-Frontend/
> > > ![image](https://user-images.githubusercontent.com/56157504/136036170-d8dbbb6e-0fca-432b-9ba8-d4685d09470a.png)
> > > This is the screenshot of the network tab of the currently deployed version. You can see the call to page =3 and page =4 are called simultaneously but API with page =4 resolves before the API with page =3 hence getData() comes out of event loop and the data with page=4 gets rendered before the data with page=3 and hence the clutter.
> > 
> > 
> > Yeah, true. but JS is single threaded and even if multiple getData() gets called, it wont execute simultaneouly. For example : refer to this codepen I made [codepen](https://codepen.io/varuntiwari2001/pen/yLXdKVd) On every scroll the getData() will be called and for every promise returned by getData(), addToTable() will be called. If users scrolls the getData() function calls will be queued. The queued functions will only resume after the call stack is empty. The getData() functions in the queue will be in order of page number, and for each data set recieved, the addToTable() will run, before the stack becomes empty. So i dont think that's a problem 🤔. I may be wrong. Pls correct me if making any mistake 😊
> 
> I edited the [codepen](https://codepen.io/batflarrow/pen/gORNzaE?editors=1011) slightly to introduce time lag like the apis each time getData() is called it waits for less amount of time to resolve so if getData() was called for the first time it will wait for 5 seconds if it was called for the second time it will wait for 4 seconds and so on and also after resolving I send the time value when the function was called and sent it to the addToPage() function. So according to you if 2-3 getData() are called simultaneously the function with time = 5 should be called first and then time =4 and then time =3 and then the div's should be printed in the order with time =5 then time =4 then time =3 but you can clearly see that's not happening, in the code pen. 
Do correct me if my code is wrong.
[image](https://user-images.githubusercontent.com/56157504/136052449-d5443ec6-5d2a-4105-ae65-b03f448fbb67.png)

**Comment by infern018** [comment]
Thanks a ton @varunKT001  @batflarrow , the main aim of OpenCode is to engage in such fruitful discussions about coding practices, even I got to learn a lot following this thread ^^

**Comment by varunKT001** [comment]
> > > > > > The issue is not with the order but with the API being called even if one call is made already. A better fix will be to only fetch the next results after the previous results have been fetched and rendered. The issue is that when scrolling down we have to just add a check if a current request is being made then not make another request. But currently what happening is multiple requests are being made because of which the ordering is getting a little bit shuffled. A single boolean variable isFetching will clear out this issue. I wrote the code for it already but forgot to push it in my pr so this one's on me.
> > > > > 
> > > > > 
> > > > > In my opinion it should not be the case. Whenever the getData() function will be called by the scroll event listner, it will go in the queue, and will begin its execution after the call stack has been empty. Therefore the new results are only fetched when the previous results are rendered.
> > > > 
> > > > 
> > > > But currently, getData() function was being called multiple times and it is an asynchronous function it was running in parallel, and each getData() call was not ending after the same period of time so if the getData() function is called simultaneously for page=4 and page=5 it may happen that the results with the page=5 will be fetched before the results with page =4 so the results on page=5 were getting rendered before the page=4 and hence the clutter. Yes, getData() function will move back to call Stack once it gets executed but here multiple getData() were being pushed onto the call stack simultaneously on reaching the end of the page. I fixed this in my pr by only calling the getData() if one getData() is already not in the queue. I had implemented this earlier itself but somehow forgot to push it in my pr and it's working fine now. https://batflarrow.github.io/Leaderboard-Frontend/
> > > > ![image](https://user-images.githubusercontent.com/56157504/136036170-d8dbbb6e-0fca-432b-9ba8-d4685d09470a.png)
> > > > This is the screenshot of the network tab of the currently deployed version. You can see the call to page =3 and page =4 are called simultaneously but API with page =4 resolves before the API with page =3 hence getData() comes out of event loop and the data with page=4 gets rendered before the data with page=3 and hence the clutter.
> > > 
> > > 
> > > Yeah, true. but JS is single threaded and even if multiple getData() gets called, it wont execute simultaneouly. For example : refer to this codepen I made [codepen](https://codepen.io/varuntiwari2001/pen/yLXdKVd) On every scroll the getData() will be called and for every promise returned by getData(), addToTable() will be called. If users scrolls the getData() function calls will be queued. The queued functions will only resume after the call stack is empty. The getData() functions in the queue will be in order of page number, and for each data set recieved, the addToTable() will run, before the stack becomes empty. So i dont think that's a problem 🤔. I may be wrong. Pls correct me if making any mistake 😊
> > 
> > 
> > I edited the [codepen](https://codepen.io/batflarrow/pen/gORNzaE?editors=1011) slightly to introduce time lag like the apis each time getData() is called it waits for less amount of time to resolve so if getData() was called for the first time it will wait for 5 seconds if it was called for the second time it will wait for 4 seconds and so on and also after resolving I send the time value when the function was called and sent it to the addToPage() function. So according to you if 2-3 getData() are called simultaneously the function with time = 5 should be called first and then time =4 and then time =3 and then the div's should be printed in the order with time =5 then time =4 then time =3 but you can clearly see that's not happening, in the code pen.
> > Do correct me if my code is wrong.
> > [image](https://user-images.githubusercontent.com/56157504/136052449-d5443ec6-5d2a-4105-ae65-b03f448fbb67.png)

Yes. I was missing the point that the response timing may vary. Thats why I also edited the comment when I realised that. Thanks BTW 😃😃

**Comment by varunKT001** [comment]
> Thanks a ton @varunKT001 @batflarrow , the main aim of OpenCode is to engage in such fruitful discussions about coding practices, even I got to learn a lot following this thread ^^

Yeah 😃, The main purpose is learning. Thanks @batflarrow @infern018

**Comment by batflarrow** [comment]
> > Thanks a ton @varunKT001 @batflarrow , the main aim of OpenCode is to engage in such fruitful discussions about coding practices, even I got to learn a lot following this thread ^^
> 
> Yeah 😃, The main purpose is learning. Thanks @batflarrow @infern018

Yeah it was fun 🎉 !!
